iroha/src/sumeragi.rs

```diff
@@ -161,68 +230,170 @@ impl Sumeragi {
+                            if *peer != self.peer_id {
+                                send_futures.push(message.clone().send_to(peer));
                             }
                         }
-                        None => self.voting_block = Some(block),
-                    };
-                    if let Some(block) = self.voting_block.clone() {
-                        if block.signatures.len() >= 2 * self.max_faults {
-                            let block = block.sign(&self.public_key, &self.private_key)?;
-                            let mut send_futures = Vec::new();
-                            for peer in self.validating_peers() {
-                                send_futures.push(Message::Committed(block.clone()).send_to(peer));
-                            }
-                            for peer in self.peers_set_b() {
-                                send_futures.push(Message::Committed(block.clone()).send_to(peer));
-                            }
-                            send_futures
-                                .push(Message::Committed(block.clone()).send_to(self.leader()));
-                            let _results = futures::future::join_all(send_futures).await;
-                            let block = block.validate(&*self.world_state_view.read().await)?;
-                            let hash = block.hash();
-                            self.blocks_sender.write().await.send(block).await;
-                            self.next_round(hash);
+                        if *self.leader() != self.peer_id {
+                            send_futures.push(message.clone().send_to(self.leader()))
                         }
+                        for peer in self.peers_set_b() {
+                            send_futures.push(message.clone().send_to(peer));
+                        }
+                        send_futures.push(message.clone().send_to(self.proxy_tail()));
+                        futures::future::join_all(send_futures).await;
                     }
                 }
             }
-            Message::Committed(block) => {
-                //TODO: check if the block is the same as pending
-                let block = block.validate(&*self.world_state_view.read().await)?;
-                let hash = block.hash();
-                self.blocks_sender.write().await.send(block).await;
-                self.next_round(hash);
-            }
+        }
+        let minimum_votes = self.max_faults + 1;
+        if commit_timeout.signatures.len() >= minimum_votes {
+            //TODO: store invalidated block hashes
+            self.change_view().await;
         }
         Ok(())
     }
 
-    fn next_round(&mut self, prev_block_hash: Hash) {
+    async fn next_round(&mut self, prev_block_hash: Hash) {
         Self::sort_peers(&mut self.sorted_peers, Some(prev_block_hash));
-        self.voting_block = None;
+        *self.voting_block.write().await = None;
+    }
+
+    async fn change_view(&mut self) {
```

Comment:
Why such a naming? 
It's may confuse and I thought it's something WSV related, but it's just another sort of peers.

Reply:
The naming is from [whitepaper](https://github.com/hyperledger/iroha/blob/iroha2-wp/iroha_2_whitepaper.md#the-basics): `if there is no commit message in time, the node tries to convince the network to do a view change, which creates a new leader and proxy tail by shifting down the 2f+1 nodes used by 1 to the right`.

The view change mentioned here is different from the sorting based on the hash of previous block. It shifts the order by one instead. I suggest to keep this naming to be consistent with whitepaper.

Reply:
not in this pr, but would be great to extract all this logic into private view struct of sumeragi

Reply:
Okay, makes sense, I will think about it.